File: node/src/scanners/skill-scanner.ts
/**
 * Optional DEEP skill-review engine — wraps the external `skill-scanner` CLI.
 *
 * Parity with `python/rafter_cli/scanners/skill_scanner.py` (bead sable-7g7).
 * This is the **couple, don't swap** integration: the zero-dependency
 * deterministic quick scan stays the default for `rafter agent audit-skill`;
 * passing `--deep` shells out to Cisco AI Defense's `skill-scanner`
 * (pip: `cisco-ai-skill-scanner`) for a deeper pass covering prompt injection,
 * taint/dataflow, YARA and .pyc integrity — the blind spots the regex quick
 * scan cannot see.
 *
 * Design mirrors `betterleaks.ts`: an external tool both runtimes shell out to
 * and whose JSON we parse. Critically, we invoke **only the offline/static
 * default analyzers** (static + bytecode + pipeline). We never pass
 * `--use-llm`, `--use-virustotal`, `--use-aidefense` or `--use-behavioral`, so
 * nothing leaves the machine — preserving Rafter's offline / no-telemetry
 * promise. The FORBIDDEN_FLAGS invariant is asserted by the vitest suite.
 *
 * Observed `skill-scanner` version: 2.0.11.
 */
import { execFile, execSync } from "child_process";
import { promisify } from "util";
import path from "path";
import fs from "fs";
import { askYesNo } from "../utils/prompt.js";

const execFileAsync = promisify(execFile);

/** PyPI package providing the external CLI, and the version we pin (mirrors
 * SKILL_SCANNER_VERSION in the Python side / BETTERLEAKS_VERSION). */
export const SKILL_SCANNER_PACKAGE = "cisco-ai-skill-scanner";
export const SKILL_SCANNER_VERSION = "2.0.11";

/** skill-scanner severity (UPPERCASE) -> our tier (lowercase). skill-scanner
 * also emits INFO, mapped to "low" (informational, e.g. missing-license). */
const SEVERITY_MAP: Record<string, string> = {
  CRITICAL: "critical",
  HIGH: "high",
  MEDIUM: "medium",
  LOW: "low",
  INFO: "low",
};

/** Severities that count as actionable findings for exit-code purposes.
 * INFO/low policy hints do NOT flip the exit code (matches the quick scan). */
const FINDING_SEVERITIES = new Set(["critical", "high", "medium"]);

/** Network/LLM/cloud flags that must NEVER appear in our argv. Enforced by a
 * test so a regression that flips on a remote analyzer fails the suite. */
export const FORBIDDEN_FLAGS = [
  "--use-llm",
  "--use-virustotal",
  "--use-aidefense",
  "--use-behavioral",
  "--vt-api-key",
  "--aidefense-api-key",
];

export const INSTALL_HINT =
  "skill-scanner not found. The --deep engine requires Cisco AI Defense's " +
  "skill-scanner. Install it with the managed installer:\n" +
  "    rafter agent update-skill-scanner\n" +
  "  (or manually: uv tool install cisco-ai-skill-scanner)\n" +
  "Then re-run with --deep.";

export interface DeepFinding {
  ruleId: string;
  severity: string; // our tier: critical/high/medium/low
  category: string;
  title: string;
  description: string;
  file: string | null;
  line: number | null;
  snippet: string | null;
  analyzer: string;
}

export interface DeepScanResult {
  available: boolean;
  findings: DeepFinding[];
  maxSeverity: string | null;
  analyzersUsed: string[];
  error: string;
  raw: Record<string, unknown> | null;
}

export function hasFindings(result: DeepScanResult): boolean {
  return result.findings.some((f) => FINDING_SEVERITIES.has(f.severity));
}

interface BuildArgvOpts {
  skillFile?: string | null;
  lenient?: boolean;
}

export class SkillScanner {
  private resolvedPath: string | null = null;

  /** Locate the `skill-scanner` launcher on PATH (uv tool / pip --user both
   * place it there). Cached after first lookup. */
  private resolvePath(): string | null {
    if (this.resolvedPath !== null) return this.resolvedPath || null;
    const cmd =
      process.platform === "win32" ? "where skill-scanner" : "which skill-scanner";
    try {
      const result = execSync(cmd, { timeout: 5000, encoding: "utf-8" });
      const found = result.trim().split("\n")[0].trim();
      this.resolvedPath = found || "";
      return found || null;
    } catch {
      this.resolvedPath = "";
      return null;
    }
  }

  isAvailable(): boolean {
    return this.resolvePath() !== null;
  }

  /**
   * Construct the OFFLINE-SAFE argv for a skill-scanner scan.
   *
   * Guarantees (asserted by tests): no flag in FORBIDDEN_FLAGS is ever added,
   * so only the default static/bytecode/pipeline analyzers run, all offline.
   * `--format json` for a machine-parseable object; `--fail-on-severity medium`
   * so the exit code reflects findings (skill-scanner otherwise exits 0 even on
   * CRITICAL).
   */
  static buildArgv(targetDir: string, opts: BuildArgvOpts = {}): string[] {
    const argv = [
      "scan",
      targetDir,
      "--format",
      "json",
      "--fail-on-severity",
      "medium",
    ];
    if (opts.skillFile) {
      argv.push("--skill-file", opts.skillFile);
    }
    if (opts.lenient) {
      argv.push("--lenient");
    }
    return argv;
  }

  /**
   * Run an offline deep scan for a skill file or directory. skill-scanner only
   * scans a *directory*, so when given a file we scan its parent directory and
   * point --skill-file at the filename (plus --lenient for robustness).
   */
  async scanPath(skillPath: string): Promise<DeepScanResult> {
    const binary = this.resolvePath();
    if (!binary) {
      return {
        available: false,
        findings: [],
        maxSeverity: null,
        analyzersUsed: [],
        error: INSTALL_HINT,
        raw: null,
      };
    }

    let targetDir: string;
    let skillFile: string | null;
    let lenient: boolean;
    const isDir = fs.existsSync(skillPath) && fs.statSync(skillPath).isDirectory();
    if (isDir) {
      targetDir = skillPath;
      skillFile = null;
      lenient = false;
    } else {
      targetDir = path.dirname(skillPath);
      skillFile = path.basename(skillPath);
      lenient = true;
    }

    const argv = SkillScanner.buildArgv(targetDir, { skillFile, lenient });

    let stdout = "";
    try {
      const res = await execFileAsync(binary, argv, {
        timeout: 120_000,
        maxBuffer: 32 * 1024 * 1024,
      });
      stdout = (res.stdout || "").trim();
    } catch (e: unknown) {
      // skill-scanner exits non-zero (1) when findings hit the --fail-on-severity
      // floor; the JSON report is still on stdout in that case. execFile rejects
      // on non-zero exit, so recover stdout from the error object.
      const err = e as { stdout?: string; killed?: boolean; signal?: string; message?: string };
      if (err.killed || err.signal === "SIGTERM") {
        return errorResult("skill-scanner scan timed out");
      }
      stdout = (err.stdout || "").trim();
      if (!stdout) {
        return errorResult(`skill-scanner invocation failed: ${err.message || e}`);
      }
    }

    if (!stdout) {
      return errorResult("skill-scanner produced no JSON output");
    }

    let parsed: Record<string, unknown>;
    try {
      parsed = JSON.parse(stdout);
    } catch (e) {
      return errorResult(`failed to parse skill-scanner JSON: ${e}`);
    }

    return SkillScanner.map(parsed);
  }

  static map(parsed: Record<string, unknown>): DeepScanResult {
    const rawFindings = Array.isArray(parsed.findings) ? parsed.findings : [];
    const findings: DeepFinding[] = rawFindings.map((f: Record<string, unknown>) => {
      const rawSev = String(f.severity ?? "").toUpperCase();
      const tier = SEVERITY_MAP[rawSev] ?? "low";
      return {
        ruleId: String(f.rule_id ?? f.id ?? "unknown"),
        severity: tier,
        category: String(f.category ?? ""),
        title: String(f.title ?? ""),
        description: String(f.description ?? ""),
        file: (f.file_path as string) ?? null,
        line: (f.line_number as number) ?? null,
        snippet: (f.snippet as string) ?? null,
        analyzer: String(f.analyzer ?? ""),
      };
    });

    const rawMax = parsed.max_severity;
    const maxSeverity = rawMax ? SEVERITY_MAP[String(rawMax).toUpperCase()] ?? null : null;
    const analyzersUsed = Array.isArray(parsed.analyzers_used)
      ? (parsed.analyzers_used as string[])
      : [];

    return {
      available: true,
      findings,
      maxSeverity,
      analyzersUsed,
      error: "",
      raw: parsed,
    };
  }
}

function errorResult(error: string): DeepScanResult {
  return {
    available: true,
    findings: [],
    maxSeverity: null,
    analyzersUsed: [],
    error,
    raw: null,
  };
}

export interface InstallResult {
  ok: boolean;
  message: string;
  via: string; // "uv" | "pip" | ""
}

/**
 * Managed installer for the optional `skill-scanner` deep engine.
 *
 * skill-scanner is a HEAVY PyPI package (litellm, fastapi, yara-x, …), so we
 * install it ISOLATED rather than into any shared environment:
 *   1. `uv tool install cisco-ai-skill-scanner==<version>` (preferred) — uv
 *      builds a dedicated venv and exposes a `skill-scanner` launcher on PATH.
 *   2. Fallback `python3 -m pip install --user cisco-ai-skill-scanner==<version>`.
 *
 * Security posture (mirrors the betterleaks installer's intent): pinned version,
 * list-form `execFile` (never a shell — no command injection), user-scoped, no
 * elevation. Integrity relies on TLS-to-PyPI + the version pin (a single-binary
 * SHA256 pin like betterleaks' is not possible over a pip transitive tree
 * without a lockfile — documented limitation, not a regression).
 */
export class SkillScannerInstaller {
  static uvPath(): string | null {
    const cmd = process.platform === "win32" ? "where uv" : "which uv";
    try {
      const result = execSync(cmd, { timeout: 5000, encoding: "utf-8" });
      const found = result.trim().split("\n")[0].trim();
      return found || null;
    } catch {
      return null;
    }
  }

  /** Construct the (list-form) install command + argv. Version is pinned with
   * `==` so it can never be read as extra arguments. */
  static buildInstall(
    version: string,
    uv: string | null,
  ): { cmd: string; argv: string[] } {
    const spec = `${SKILL_SCANNER_PACKAGE}==${version}`;
    if (uv) {
      return { cmd: uv, argv: ["tool", "install", "--force", spec] };
    }
    return {
      cmd: process.platform === "win32" ? "python" : "python3",
      argv: ["-m", "pip", "install", "--user", "--upgrade", spec],
    };
  }

  async install(
    version: string = SKILL_SCANNER_VERSION,
    onProgress?: (msg: string) => void,
  ): Promise<InstallResult> {
    const uv = SkillScannerInstaller.uvPath();
    const via = uv ? "uv" : "pip";
    const { cmd, argv } = SkillScannerInstaller.buildInstall(version, uv);
    if (onProgress) {
      onProgress(`Installing ${SKILL_SCANNER_PACKAGE}==${version} via ${via}…`);
    }
    try {
      await execFileAsync(cmd, argv, {
        timeout: 900_000, // heavy transitive tree; allow generous build time
        maxBuffer: 32 * 1024 * 1024,
      });
    } catch (e: unknown) {
      const err = e as { stderr?: string; stdout?: string; message?: string };
      const tail = (err.stderr || err.stdout || err.message || String(e)).trim().slice(-800);
      return { ok: false, message: `installer failed: ${tail || "(no output)"}`, via };
    }

    // Verify the launcher is now reachable.
    const checkCmd =
      process.platform === "win32" ? "where skill-scanner" : "which skill-scanner";
    try {
      const result = execSync(checkCmd, { timeout: 5000, encoding: "utf-8" });
      const found = result.trim().split("\n")[0].trim();
      if (found) return { ok: true, message: found, via };
    } catch {
      /* fall through */
    }
    return {
      ok: false,
      message:
        "install reported success but `skill-scanner` is not on PATH. If you " +
        "used the pip fallback, ensure your user-site bin directory is on PATH.",
      via,
    };
  }
}

/** Severity tiers, low→high, used to escalate a report's severity by deep
 * findings. Mirrors the order used by skill review. */
const _TIER_ORDER = ["clean", "low", "medium", "high", "critical"];

/**
 * The highest **actionable** tier (medium/high/critical) among a deep result's
 * findings, or "clean". low/INFO findings are reported but never escalate the
 * overall severity / exit code — matching the quick-scan contract.
 */
export function deepSeverityTier(result: DeepScanResult): string {
  let tier = "clean";
  for (const f of result.findings) {
    if (FINDING_SEVERITIES.has(f.severity) && _TIER_ORDER.indexOf(f.severity) > _TIER_ORDER.indexOf(tier)) {
      tier = f.severity;
    }
  }
  return tier;
}

/** Count of actionable (medium+) deep findings. */
export function deepActionableCount(result: DeepScanResult): number {
  return result.findings.filter((f) => FINDING_SEVERITIES.has(f.severity)).length;
}

/**
 * Resolve a usable SkillScanner for an opt-in --deep run, making it **easy**:
 * if the engine isn't installed and we're on an interactive TTY (and not in
 * --json mode), offer to install it in place. Returns a ready scanner, or null
 * when it's unavailable and the caller should print the install hint + exit 2.
 */
export async function ensureSkillScanner(opts: { json?: boolean } = {}): Promise<SkillScanner | null> {
  let scanner = new SkillScanner();
  if (scanner.isAvailable()) return scanner;

  const interactive = !!process.stdin.isTTY && !opts.json;
  if (interactive) {
    process.stderr.write("\nThe --deep engine (skill-scanner) is not installed.\n");
    const yes = await askYesNo(
      "Install it now? (heavy third-party package, isolated via uv/pip)",
      false,
    );
    if (yes) {
      const result = await new SkillScannerInstaller().install(
        SKILL_SCANNER_VERSION,
        (m) => process.stderr.write(`  ${m}\n`),
      );
      if (result.ok) {
        scanner = new SkillScanner();
        if (scanner.isAvailable()) {
          process.stderr.write(`skill-scanner installed (${result.via}).\n`);
          return scanner;
        }
      } else {
        process.stderr.write(`Install failed: ${result.message}\n`);
      }
    }
  }
  return null;
}

File: python/rafter_cli/scanners/skill_scanner.py
"""Optional DEEP skill-review engine — wraps the external `skill-scanner` CLI.

PoC (bead sable-7g7). This is the **couple, don't swap** integration: our
zero-dependency deterministic quick scan stays the default for
``rafter audit-skill``; passing ``--deep`` shells out to Cisco AI Defense's
``skill-scanner`` (pip: ``cisco-ai-skill-scanner``) for a deeper pass that
covers prompt injection, taint/dataflow, YARA and .pyc integrity — the blind
spots our regex quick scan cannot see.

Design mirrors ``betterleaks.py``: an external tool both runtimes shell out to
and whose JSON we parse. Critically, we invoke **only the offline/static
default analyzers** (static + bytecode + pipeline). We never pass
``--use-llm``, ``--use-virustotal``, ``--use-aidefense`` (or behavioral, which
is static but kept off for the PoC to stay minimal), so nothing leaves the
machine — preserving Rafter's offline / no-telemetry promise.

Observed ``skill-scanner`` version: 2.0.11.
"""
from __future__ import annotations

import json
import shutil
import subprocess
import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

# PyPI package providing the external `skill-scanner` CLI, and the version we
# pin for reproducibility. skill-scanner's JSON shape is stable but not a
# documented contract, so the managed installer pins this exact version (mirrors
# BETTERLEAKS_VERSION). Bump deliberately after re-validating the JSON mapping.
SKILL_SCANNER_PACKAGE = "cisco-ai-skill-scanner"
SKILL_SCANNER_VERSION = "2.0.11"

# skill-scanner severity (UPPERCASE) -> our tier (lowercase).
# Our tiers: critical / high / medium / low. skill-scanner also emits INFO,
# which we map to "low" (informational, e.g. missing-license policy hints).
_SEVERITY_MAP: dict[str, str] = {
    "CRITICAL": "critical",
    "HIGH": "high",
    "MEDIUM": "medium",
    "LOW": "low",
    "INFO": "low",
}

# Severities that count as actionable "findings" for exit-code purposes.
# INFO/low policy hints (e.g. missing license) do NOT flip the exit code,
# matching the spirit of our quick scan (secrets / high-risk commands only).
_FINDING_SEVERITIES = frozenset({"critical", "high", "medium"})

INSTALL_HINT = (
    "skill-scanner not found. The --deep engine requires Cisco AI Defense's "
    "skill-scanner. Install it with the managed installer:\n"
    "    rafter agent update-skill-scanner\n"
    "  (or manually: uv tool install cisco-ai-skill-scanner)\n"
    "Then re-run with --deep."
)


@dataclass
class DeepFinding:
    """One mapped finding from skill-scanner, in our normalized shape."""
    rule_id: str
    severity: str  # our tier: critical/high/medium/low
    category: str
    title: str
    description: str
    file_path: str | None
    line: int | None
    snippet: str | None
    analyzer: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "ruleId": self.rule_id,
            "severity": self.severity,
            "category": self.category,
            "title": self.title,
            "description": self.description,
            "file": self.file_path,
            "line": self.line,
            "snippet": self.snippet,
            "analyzer": self.analyzer,
        }


@dataclass
class DeepScanResult:
    available: bool
    findings: list[DeepFinding] = field(default_factory=list)
    max_severity: str | None = None  # our tier or None
    analyzers_used: list[str] = field(default_factory=list)
    error: str = ""  # populated when available is False or scan failed
    raw: dict[str, Any] | None = None  # raw skill-scanner JSON (for --json passthrough)

    @property
    def has_findings(self) -> bool:
        """True if any finding is at/above the actionable severity floor."""
        return any(f.severity in _FINDING_SEVERITIES for f in self.findings)


class SkillScanner:
    """Thin wrapper around the external `skill-scanner` CLI (offline analyzers only)."""

    def __init__(self) -> None:
        self._path: str | None = shutil.which("skill-scanner")

    def is_available(self) -> bool:
        return self._path is not None

    @staticmethod
    def build_argv(
        target_dir: str,
        *,
        binary: str = "skill-scanner",
        skill_file: str | None = None,
        lenient: bool = False,
    ) -> list[str]:
        """Construct the OFFLINE-SAFE argv for a skill-scanner scan.

        Guarantees (asserted by tests): NO network/LLM/cloud flags are ever
        added — no --use-llm, --use-virustotal, --use-aidefense, --use-behavioral.
        Only the default static/bytecode/pipeline analyzers run, all offline.

        We force JSON output and use --fail-on-severity so the process exit
        code reflects findings (skill-scanner otherwise exits 0 even when it
        flags critical issues).
        """
        argv = [
            binary,
            "scan",
            target_dir,
            "--format",
            "json",
            # Exit non-zero when a medium+ finding exists, so our wrapper can
            # corroborate parsed results against the process exit code.
            "--fail-on-severity",
            "medium",
        ]
        if skill_file:
            # Point skill-scanner at a non-SKILL.md metadata file (our
            # audit-skill takes an arbitrary .md path).
            argv += ["--skill-file", skill_file]
        if lenient:
            # Tolerate malformed/Claude-command-style skills.
            argv += ["--lenient"]
        return argv

    def scan_path(self, skill_path: str) -> DeepScanResult:
        """Run an offline deep scan for a skill file or directory.

        skill-scanner operates on a *directory*. ``audit-skill`` receives a
        file path, so when given a file we scan its parent directory and point
        ``--skill-file`` at the filename (plus ``--lenient`` for robustness).
        """
        if not self._path:
            return DeepScanResult(available=False, error=INSTALL_HINT)

        p = Path(skill_path)
        if p.is_dir():
            target_dir = str(p)
            skill_file = None
            lenient = False
        else:
            target_dir = str(p.parent)
            skill_file = p.name
            lenient = True

        argv = self.build_argv(
            target_dir,
            binary=self._path,
            skill_file=skill_file,
            lenient=lenient,
        )

        try:
            result = subprocess.run(
                argv,
                capture_output=True,
                text=True,
                timeout=120,
            )
        except subprocess.TimeoutExpired:
            return DeepScanResult(available=True, error="skill-scanner scan timed out")
        except (OSError, FileNotFoundError) as exc:
            return DeepScanResult(available=True, error=f"skill-scanner invocation failed: {exc}")

        # skill-scanner exit codes: 0 = clean OR (findings present but below
        # --fail-on-severity floor); 1 = findings at/above floor; 2 = usage
        # error. The JSON report is on stdout regardless. Parse it; only treat
        # a missing/invalid report as an error.
        stdout = (result.stdout or "").strip()
        if not stdout:
            stderr_tail = (result.stderr or "").strip()[-500:]
            return DeepScanResult(
                available=True,
                error=f"skill-scanner produced no JSON (exit {result.returncode}): {stderr_tail or '(no stderr)'}",
            )

        try:
            parsed = json.loads(stdout)
        except json.JSONDecodeError as exc:
            return DeepScanResult(available=True, error=f"failed to parse skill-scanner JSON: {exc}")

        return self._map(parsed)

    @staticmethod
    def _map(parsed: dict[str, Any]) -> DeepScanResult:
        findings: list[DeepFinding] = []
        for f in parsed.get("findings", []) or []:
            raw_sev = str(f.get("severity", "")).upper()
            tier = _SEVERITY_MAP.get(raw_sev, "low")
            findings.append(
                DeepFinding(
                    rule_id=str(f.get("rule_id") or f.get("id") or "unknown"),
                    severity=tier,
                    category=str(f.get("category") or ""),
                    title=str(f.get("title") or ""),
                    description=str(f.get("description") or ""),
                    file_path=f.get("file_path"),
                    line=f.get("line_number"),
                    snippet=f.get("snippet"),
                    analyzer=str(f.get("analyzer") or ""),
                )
            )

        raw_max = parsed.get("max_severity")
        max_tier = _SEVERITY_MAP.get(str(raw_max).upper()) if raw_max else None

        return DeepScanResult(
            available=True,
            findings=findings,
            max_severity=max_tier,
            analyzers_used=list(parsed.get("analyzers_used") or []),
            raw=parsed,
        )


@dataclass
class InstallResult:
    ok: bool
    message: str
    via: str = ""  # "uv" | "pip" | ""


class SkillScannerInstaller:
    """Managed installer for the optional `skill-scanner` deep engine.

    skill-scanner is a HEAVY PyPI package (it pulls litellm, fastapi, yara-x,
    tokenizers, …), so — unlike a hard dependency — we install it in an
    **isolated** environment that cannot perturb Rafter's own dependency tree:

    1. ``uv tool install cisco-ai-skill-scanner==<version>`` (preferred): uv
       builds a dedicated venv and exposes a ``skill-scanner`` launcher on PATH.
    2. Fallback ``python -m pip install --user cisco-ai-skill-scanner==<version>``
       when uv is absent.

    Security posture (mirrors the betterleaks installer's intent):
    - **Pinned version** for reproducibility (``SKILL_SCANNER_VERSION``).
    - **List-form subprocess**, never ``shell=True`` — no command injection.
    - No elevated privileges; user-scoped install only.
    - Integrity relies on TLS-to-PyPI + the version pin. (Unlike the betterleaks
      single-binary download we cannot pin a SHA256 over the whole transitive
      tree without a lockfile; this is a documented limitation, not a regression.)

    This only *installs* the engine. It does not change the offline-only
    invocation contract enforced by ``SkillScanner.build_argv``.
    """

    @staticmethod
    def uv_path() -> str | None:
        return shutil.which("uv")

    @staticmethod
    def build_install_argv(version: str, *, uv: str | None) -> list[str]:
        """Construct the (list-form) install argv. Version is pinned with ``==``.

        ``version`` must be a plain version string; we never interpolate it into
        a shell, and the ``==`` pin prevents it from being read as extra args.
        """
        spec = f"{SKILL_SCANNER_PACKAGE}=={version}"
        if uv:
            # --force so an existing managed install is replaced (update semantics).
            return [uv, "tool", "install", "--force", spec]
        # Fallback: user-site pip install via the running interpreter.
        py = sys.executable or "python3"
        return [py, "-m", "pip", "install", "--user", "--upgrade", spec]

    def install(
        self,
        version: str = SKILL_SCANNER_VERSION,
        on_progress=None,
    ) -> InstallResult:
        uv = self.uv_path()
        argv = self.build_install_argv(version, uv=uv)
        via = "uv" if uv else "pip"
        if on_progress:
            on_progress(f"Installing {SKILL_SCANNER_PACKAGE}=={version} via {via}…")
        try:
            result = subprocess.run(
                argv,
                capture_output=True,
                text=True,
                timeout=900,  # heavy transitive tree; allow generous build time
            )
        except subprocess.TimeoutExpired:
            return InstallResult(False, "skill-scanner install timed out", via)
        except (OSError, FileNotFoundError) as exc:
            return InstallResult(False, f"install invocation failed: {exc}", via)

        if result.returncode != 0:
            tail = (result.stderr or result.stdout or "").strip()[-800:]
            return InstallResult(
                False,
                f"installer exited {result.returncode}: {tail or '(no output)'}",
                via,
            )

        # Verify the launcher is now reachable and runnable.
        path = shutil.which("skill-scanner")
        if not path:
            return InstallResult(
                False,
                "install reported success but `skill-scanner` is not on PATH. "
                "If you used the pip fallback, ensure your user-site bin "
                "directory is on PATH.",
                via,
            )
        return InstallResult(True, path, via)

    @staticmethod
    def build_uninstall_argv(*, uv: str | None) -> list[str]:
        """The (list-form) uninstall argv. Mirrors install: uv tool, pip fallback."""
        if uv:
            return [uv, "tool", "uninstall", SKILL_SCANNER_PACKAGE]
        py = sys.executable or "python3"
        return [py, "-m", "pip", "uninstall", "-y", SKILL_SCANNER_PACKAGE]

    def uninstall(self, on_progress=None) -> InstallResult:
        """Remove the managed skill-scanner. Idempotent: a no-op (success) when
        it isn't installed. Tries `uv tool uninstall` first (how `install`
        prefers to set it up), then a `pip uninstall` fallback — because we don't
        durably record which path installed it."""
        if not shutil.which("skill-scanner"):
            return InstallResult(True, "skill-scanner is not installed (nothing to do)", "")

        attempts: list[str] = []
        uv = self.uv_path()
        methods: list[tuple[str, list[str]]] = []
        if uv:
            methods.append(("uv", self.build_uninstall_argv(uv=uv)))
        methods.append(("pip", self.build_uninstall_argv(uv=None)))
        for via, argv in methods:
            if on_progress:
                on_progress(f"Removing skill-scanner via {via}…")
            try:
                result = subprocess.run(
                    argv, capture_output=True, text=True, timeout=120
                )
                attempts.append(f"{via}:{result.returncode}")
            except (subprocess.TimeoutExpired, OSError, FileNotFoundError) as exc:
                attempts.append(f"{via}:err({exc})")
            if not shutil.which("skill-scanner"):
                return InstallResult(True, f"removed via {via}", via)

        return InstallResult(
            False,
            f"skill-scanner is still on PATH after uninstall attempts ({', '.join(attempts)}). "
            "It may have been installed by another tool; remove it manually.",
            "",
        )


# Severity tiers, low→high — used to escalate a report's severity by deep findings.
_TIER_ORDER: tuple[str, ...] = ("clean", "low", "medium", "high", "critical")


def deep_severity_tier(result: DeepScanResult) -> str:
    """Highest **actionable** tier (medium/high/critical) among findings, or
    'clean'. low/INFO findings never escalate the overall severity / exit code,
    matching the quick-scan contract."""
    tier = "clean"
    for f in result.findings:
        if f.severity in _FINDING_SEVERITIES and _TIER_ORDER.index(f.severity) > _TIER_ORDER.index(tier):
            tier = f.severity
    return tier


def deep_actionable_count(result: DeepScanResult) -> int:
    """Count of actionable (medium+) deep findings."""
    return sum(1 for f in result.findings if f.severity in _FINDING_SEVERITIES)


def ensure_skill_scanner(*, json_out: bool = False) -> "SkillScanner | None":
    """Resolve a usable SkillScanner for an opt-in --deep run, making it **easy**:
    if the engine isn't installed and we're on an interactive TTY (and not in
    --json mode), offer to install it in place. Returns a ready scanner, or None
    when unavailable and the caller should print the install hint + exit 2."""
    scanner = SkillScanner()
    if scanner.is_available():
        return scanner

    interactive = sys.stdin.isatty() and not json_out
    if interactive:
        print("\nThe --deep engine (skill-scanner) is not installed.", file=sys.stderr)
        # Prompt on stderr to keep stdout clean.
        print(
            "  Install it now? (heavy third-party package, isolated via uv/pip) [y/N] ",
            end="",
            file=sys.stderr,
            flush=True,
        )
        try:
            ans = input().strip().lower()
        except EOFError:
            ans = ""
        if ans in ("y", "yes"):
            result = SkillScannerInstaller().install(
                on_progress=lambda m: print(f"  {m}", file=sys.stderr)
            )
            if result.ok:
                s2 = SkillScanner()
                if s2.is_available():
                    print(f"skill-scanner installed ({result.via}).", file=sys.stderr)
                    return s2
            else:
                print(f"Install failed: {result.message}", file=sys.stderr)
    return None
